daceml/autodiff/backward_pass_generator.py

```diff
@@ -0,0 +1,1270 @@
+                    "When the forward sdfg is the same as the backward sdfg, outputs must be a"
+                    "single scalar")
+            self.separate_sdfgs = False
+        else:
+            self.separate_sdfgs = True
+
+    def _expand_nodes(self, subgraph: dstate.StateSubgraphView) -> bool:
+        """ Expand all library nodes in the graph to pure implementations. Returns whether something was expanded
+        """
+
+        expanded_something = False
+        for node, state in subgraph.all_nodes_recursive():
+            if isinstance(state, dstate.StateSubgraphView):
+                state = state.graph
+
+            # check if the node exists in the backward implementation repository
+            if find_backward_implementation(state.parent, state,
+                                            node) is not None:
+                continue
+
+            # only check others if we didn't break out of the above loop
+            if isinstance(node, ONNXOp):
+                for impl in ONNXForward.registered_implementations(
+                        node.schema.name):
+                    if impl.forward_can_be_applied(node, state, self.sdfg):
+                        # try to apply the expansion
+                        class Expansion(xf.ExpandTransformation):
+                            environments = []
+                            _expansion_result = None
+
+                            @classmethod
+                            def expansion(cls, node, state, sdfg):
+                                return impl.forward(node, state, sdfg)
+
+                            @staticmethod
+                            def annotates_memlets() -> bool:
+                                return True
+
+                        Expansion._match_node = xf.PatternNode(type(node))
+                        Expansion.apply_to(state.parent,
+                                           verify=False,
+                                           _match_node=node)
+                        expanded_something = True
+                        continue
+
+            # This could later on be changed to check if the expansion is differentiable and if not, move
+            # on to the next expansion. For now we will just apply the first one that matches, prioritizing ones that have
+            # "pure" in the name
+            if isinstance(node,
+                          nd.LibraryNode) and not isinstance(node, ONNXOp):
+                # try to select an expansion
+                if hasattr(node, "implementations"):
+                    implementations = node.implementations
+
+                    def contains_pure(name, impl):
+                        return "pure" in name.lower(
+                        ) or "pure" in impl.__name__.lower()
+
+                    pure_candidates = [
+                        name for name, impl in implementations.items()
+                        if contains_pure(name, impl)
+                    ]
```

Comment:
why not ask for `"pure" in implementations`?

Reply:
Yeah seems like a better idea, I was probably thinking of some future thing at the time but it seems better to address this when the time comes